### PR TITLE
Add gene panel selection pipeline

### PR DESCRIPTION
**Pull request by Rwin2**
## Summary
- Gene panel selection skill (Steps 0-6) and toolset (SpaPROS, Random Forest, scGeneFit)
- Automatic dataset retrieval from CELLxGENE Census when no dataset is provided
- GPS MODE LOCK in leader to delegate the full pipeline to analysis_expert
- Works for any biological context (not tied to specific tissues or diseases)

## Files changed
- `gene_panel_selection.md` — new skill file with the full workflow
- `gene_panel_selection_tool.py` — new toolset with three selection methods
- `analysis_expert.md` — added GPS workflow section and dataset search instructions
- `leader.md` — added GPS MODE LOCK delegation logic
- `SKILL.md` — added reference to the GPS skill
- `__init__.py` — registered the GPS toolset

### CODE REVIEW

**Review by JiwaniZakir — COMMENTED** [review]
The 230-line workflow embedded directly in `analysis_expert.md` duplicates content that already lives in `.pantheon/skills/omics/gene_panel_selection.md` — the agent is explicitly told to re-read that skill file "before each major step," yet the full procedure is also inlined here. This creates two sources of truth: if the skill file is updated (e.g., adding a new method or changing the ARI threshold), the template won't reflect it, and the agent will receive conflicting instructions. The workflow content should either live entirely in the skill file (with the template just referencing it) or the skill file should be the authoritative source and this inline copy removed.

The hardcoded magic numbers scattered through the prompt — `max_constraints <= 1000` for `select_scgenefit`, `n_hvg < 3000` for `select_spapros`, the `>5%` ARI degradation threshold in Phase 2 — are invisible to callers and can't be overridden per-invocation. If these constraints need tuning for a specific dataset, the only option is editing the template. Surfacing them as parameters in the toolset schema or as explicit leader-context fields would make the pipeline more flexible and testable.

There's also a typo in the modified line around line 144: `"revelant skill"` should be `"relevant skill"`, which is minor but worth fixing given this text is part of the agent's behavioral contract.

**Review by Nanguage — CHANGES_REQUESTED** [review]
Thanks for the follow-up commit! The typo fix and hyperparameter variable names are good improvements. However, several issues from the previous review remain unresolved, and there's a new one:

## Unresolved from previous review

### 1. Duplicated workflow — still two sources of truth

The ~230-line workflow inlined in `analysis_expert.md` still duplicates `gene_panel_selection.md` almost verbatim. This was the core issue raised by @JiwaniZakir. The second commit didn't address it — both files were updated in parallel, which actually makes the dual-maintenance problem worse.

**Suggestion**: Remove the inline workflow from `analysis_expert.md` and replace it with a short reference like:
```
When performing gene panel selection, **strictly** follow the workflow in 
`.pantheon/skills/omics/gene_panel_selection.md`. Re-read it before each major step.
```

### 2. Hyperparameters are named but not truly configurable

The magic numbers were replaced with variable names (`SCGENEFIT_MAX_CONSTRAINTS`, etc.) in the prompt text, but they're still hardcoded — there's no mechanism for the user or leader to override them. They exist only as a markdown table in the prompt.

## New issue introduced

### 3. Stray character in `leader.md`

Line 193 in the diff has a stray `y` character:
```
- Benchmarking section with:
+y
  - dataset splitting strategy
```

This appears to be an accidental keypress and will show up in the leader agent's prompt.

## Additional suggestions (lower priority)

- **`unwrap_llm_dict_call` decorator**: This works around a framework-level concern and shouldn't live in a domain toolset. If dict-unwrapping is needed, it belongs in the `ToolSet` base class.
- **String-typed parameters** in the toolset (`num_markers: str = "100"`) — use native types (`int`, `bool`) and let the framework handle conversion.
- **Missing graceful degradation** when `scanpy`/`spapros`/`scgenefit` are not installed — a try/import with a clear error message would be better than a raw `ImportError`.

**Comment by Nanguage** [comment]
Is it possible to remove the newly added Toolsets, or move them under "skill"? We would prefer to rely more on the built-in, general-purpose Toolsets for code execution.

**Comment by JiwaniZakir** [comment]
The GPS MODE LOCK delegation pattern in `leader.md` looks like a solid architectural choice for keeping the pipeline encapsulated in `analysis_expert`, but worth verifying that the lock logic handles edge cases where a user explicitly provides a dataset — the automatic CELLxGENE Census retrieval path shouldn't override explicit inputs. Also, having three selection methods (SpaPROS, Random Forest, scGeneFit) without a clear fallback priority or ensemble strategy in `gene_panel_selection_tool.py` may lead to inconsistent results depending on which method gets invoked. Does the skill file define criteria for method selection, or is that left to the expert agent's discretion?

**Comment by Rwin2** [comment]
Thank you for the detailed comments!

For the "two sources of ground truth" remark — this was enforced mainly because when we defined gene panel selection purely as a skill, the agent was not reliably following the pipeline and was taking too many liberties.

In the agent prompt, we explicitly encourage some level of exploration instead of blindly following a skill. That works well in general, but in this case gene panel selection is a very sensitive pipeline. Too much freedom from a generalist model (GPT, etc.) quickly leads to random search behavior, skipped steps, or poorly validated panels.

So what we did is add high-level guidelines in the agent prompt (much less detailed than the skill file). The goal is to guide the agent through the correct sequence of steps, while still allowing some flexibility. The agent prompt defines **what** needs to be done (the pipeline structure), and the skill file defines **how** to do it (details, thresholds, code, etc.).

From experience, this was the only reliable way to enforce the full pipeline without the agent deviating. Another option would have been to introduce a dedicated gene panel selection agent with a strict prompt, but we chose to keep a single reusable \`analysis_expert\` instead.

Pantheon is designed to be flexible, so alternative structures are totally welcome — we're mainly trying to conserve a general code architecture here!

**Comment by JiwaniZakir** [comment]
The GPS MODE LOCK delegation pattern in `leader.md` is worth scrutinizing — if the leader unconditionally hands off the full pipeline to `analysis_expert` without validating inputs first, you risk silent failures when CELLxGENE Census retrieval returns incomplete or mismatched datasets for the biological context. It would be worth confirming whether `gene_panel_selection_tool.py` has fallback handling when the auto-retrieved dataset doesn't meet the feature requirements for SpaPROS or scGeneFit. Also, since the three selection methods (SpaPROS, Random Forest, scGeneFit) can produce divergent panel outputs, is there a consensus or ranking step in Step 6, or does the skill leave that decision to the caller?

**Comment by JiwaniZakir** [comment]
The GPS MODE LOCK delegation pattern in `leader.md` is a solid approach, but worth confirming that the fallback behavior is well-defined if `analysis_expert` fails mid-pipeline — particularly between Steps 3 and 4 where the dataset retrieval from CELLxGENE Census could time out or return incomplete results. The three-method toolset (SpaPROS, Random Forest, scGeneFit) covering different selection paradigms is good, but the `gene_panel_selection_tool.py` should document which method is recommended as default and under what conditions each should be preferred. Also, registering the toolset in `__init__.py` — is there a version pin on the CELLxGENE Census dependency, since the API has had breaking changes across minor versions?

**Comment by Rwin2** [comment]
Re-requesting review — all points addressed on branch `gps-clean` at `0b488b8`. Quick replies below:

## Reply to @JiwaniZakir
- **N1 (230-line duplicate workflow):** defended separately — the inlined workflow in `analysis_expert.md` is intentional so the agent has it in active context; `gene_panel_selection.md` stays the canonical re-readable source. See my earlier comment on the thread.
- **N2 (hardcoded magic numbers):** fixed — introduced `GenePanelConfig` (loaded via `from_settings()`) and a new `gene_panel` block in `settings.json` with 9 tunables (`scgenefit_max_constraints`, `spapros_n_hvg`, `rf_n_estimators`, `ari_drop_threshold`, `downsample_max_cells`, `gene_count_threshold`, `split_cell_limit`, `spapros_runtime_warning_minutes`, `spapros_runtime_skip_minutes`). Nothing is hardcoded in the prompt anymore. → `5928f0d`, `d9e8c60`
- **N3 (typo "revelant"):** fixed. → `038baa3`

## Reply to @Nanguage
- **#1 (duplicated workflow — unresolved):** same response as N1 above — kept by design, canonical source is `gene_panel_selection.md`.
- **#2 (hyperparameters not truly configurable):** fixed — `GenePanelConfig.from_settings()` reads the `gene_panel` block from `settings.json`, so any user/project can override per-deployment without touching the prompt. → `5928f0d`, `d9e8c60`
- **#3 (stray `y` in `leader.md:193`):** fixed. → `038baa3`
- **S1 (`unwrap_llm_dict_call` decorator):** fixed — the whole `GenePanelToolSet` was refactored out into a plain Python library (`pantheon.toolsets.gene_panel`) with sync functions and no decorator. → `7172fdf`, `2d92866`, `d9e8c60`
- **S2 (string-typed params `num_markers: str = "100"`):** fixed — native `int` types throughout the new library. → `452619e`, `d9e8c60`
- **S3 (missing graceful degradation for optional deps):** fixed — each algorithm wraps scanpy/spapros/scgenefit imports in `try/except ImportError` and returns a dict with a clear `pip install ...` hint instead of raising. → `452619e`, `d9e8c60`

## uncovered while smoke-testing
- `f9e990b` / `0b488b8` — fix `[Errno 7] E2BIG` on kernel launch in long sessions (accumulated `toolu_*` tool-call cache in `PANTHEON_CONTEXT` exceeded Linux's `execve` 128 KB limit).
- `bd032f1` / `d9e8c60` — added a leader-owned SpaPROS runtime gate: `estimate_spapros_runtime()` returns a severity tier, and the leader calls `notify_user` with a Run/Skip choice whenever severity is `slow` or `very_slow`.

## Smoke test (end-to-end validation on this branch)
Ran the full pipeline on a 10k-cell / 22,781-gene / 56-cell-type human cancer scRNA-seq dataset with a 1000-gene immune-oncology target:
- Final panel: 1000 genes (4.4 % of transcriptome), composition 300 DE seed + 468 biologically curated (8 categories) + 232 consensus fill
- 5 test splits: ARI +3.0 %, NMI +0.9 %, silhouette +5.4 % vs full transcriptome
- Runtime ~5 h, cost ~$110. Report PDF attached in `projects/smoke_test/report.pdf` locally (not committed).

**Comment by Rwin2** [comment]
Thanks @JiwaniZakir — grouping my answers across your three comments.

**(1) Explicit user dataset vs. auto-retrieval.** `leader.md:130-141` gates this: if the user provides an AnnData path, `analysis_expert` skips CELLxGENE retrieval entirely. Auto-retrieval only fires when no dataset is passed.

**(2) Method selection / ranking.** By design there's no hardcoded priority, since each method can happen to be better than the others depending on the context — `gene_panel_selection.md` Step 3 runs each method (HVG, DE, RF, scGeneFit, SpaPROS) through an ARI-vs-K sweep on the training split and picks `(method, K*)` with the best ARI as the seed. This is just Pantheon's default mode — the user can specify exactly what they want and Pantheon will follow along rather than running all methods.

**(3) Divergent outputs.** Step 4.4 ("Consensus Fill") normalizes scores across methods and aggregates them — used as a last-resort gap filler after biological curation.

**(4) Mid-pipeline failure.** Each algorithmic method writes a scores CSV to `workdir/gene_panels/{method}/` as it completes, so a restart picks up from the last finished method without recomputing. The Pantheon agent loop handles transient retries, and Pantheon knows how to restart from preexisting results.

**Comment by JiwaniZakir** [comment]
Good point — the three methods (SpaPROS, Random Forest, scGeneFit) can be refactored as skill steps that invoke code execution through the existing general-purpose toolset rather than registering a dedicated toolset. I'll move the logic into `gene_panel_selection.md` as executable code blocks and remove `gene_panel_selection_tool.py` and the `__init__.py` registration. That should align with your preference while keeping the pipeline intact.

**Comment by Nanguage** [comment]
Superseded by #95 which contains all changes from this PR, rebased cleanly on current main.
